src/norm/mongodb.nim

```diff
@@ -0,0 +1,727 @@
+        objs: var seq[object],
+        nLimit: int,
+        offset = 0,
+        cond = %*{},
+        sort = %*{}
+      ) {.used.} =
+        objs = type(objs[^1]).getMany(nLimit, offset, cond, sort)  # this works even on an empty list; which is surprising!
+
+
+      proc update(obj: object, force = false): bool {.used, discardable.} =
+        ##[ Update document with matching ``_id`` with object field values.
+
+        ``force`` parameter ignored.
+
+        returns true if update was successful.
+        ]##
+
+        let
+          dbCollection = dbConn[customDatabase][getCollectionName(type(obj))]
+        var doc = obj.toBson()
+        let response = dbCollection.update(%*{"_id": obj.id}, doc, false, false)
+        if response.n == 1:
+          return true
+        return false
+
+      proc delete(obj: var object): bool {.used, discardable.} =
+        ## Delete a record in DB by object's id. The id is set to all zeroes after the deletion.
+
+        let
+          dbCollection = dbConn[customDatabase][getCollectionName(type(obj))]
+        let response = dbCollection.remove(%*{"_id": obj.id}, 1)
+        obj.id = Oid()
+        if response.n == 1:
+          return true
+        return false
+
+      try:
+        body
+      finally: discard
+
+  template withDb*(body: untyped): untyped {.dirty.} =
+    ##[ A wrapper for actions that require DB connection. Defines CRUD procs to work with the DB,
+    as well as ``createTables`` and ``dropTables`` procs.
+
+      Aforementioned procs and procs defined in a ``db`` block can be used only
+      in  a ``withDb`` block.
+    ]##
+
+    withCustomDb(normConnection, normUser, normPassword, normDatabase):
+      body
+
+
+proc ensureIdFields(typeSection: NimNode): NimNode =
+  ## Check if ``id`` field is in the object definition, insert it if it's not.
+
+  result = newNimNode(nnkTypeSection)
+
+  for typeDef in typeSection:
+    var objRepr = typeDef.toObjRepr()
+
+    if "id" notin objRepr.fieldNames:
+      let idField = FieldRepr(
+        signature: SignatureRepr(
+          name: "id",
+          exported: true,
+          pragmas: @[
+            PragmaRepr(name: "pk", kind: pkFlag),
+            PragmaRepr(name: "ro", kind: pkFlag),
+            PragmaRepr(name: "dbCol", kind: pkKval, value: parseExpr("\"_id\""))
+          ]
+        ),
+        typ: ident "Oid"
+      )
+      objRepr.fields.insert(idField, 0)
+
+    result.add objRepr.toTypeDef()
+
+proc reconstructType(n: NimNode): string =
+  if n.kind == nnkIdent:
+    return $n
+  if n.kind == nnkBracketExpr:
+    return "$1[$2]".format($n[0], $n[1])
+  if n.kind == nnkSym:
+    return $n
+  return "unknown"
+
+proc updateObjectRegistry(dbObjReprs: seq[ObjRepr]) {.compileTime.} =
+  var objName: string
+  for obj in dbObjReprs:
+    objName = obj.signature.name
+    if not normObjectNamesRegistry.contains(objName):
+      normObjectNamesRegistry.add objName
+
+proc genObjectToBSON(dbObjReprs: seq[ObjRepr]): string =

```

Comment:
Nim's style guide suggests Bson instead of BSON.

Reply:
Will change.